Addcourse

mayankpahwa opened the pull request.

Commit:
Added "Add Course"

 - Added adding of course
 - Added validation for date and duration of course
 - Added saving of data in JSON format.
 - GUI

Commit:
Refactored directory structure

Commit:
Added YYYY-MM-DD indicator to the date input field

mayankpahwa committed:
Add YYYY-MM-DD indicator

Add YYYY-MM-DD indicator to the Start Date field

Commit:
Removed unwanted functions that will not be used as of now

deobald commented on src/database/core.clj:
```diff
@@ -0,0 +1,23 @@
+(def course (atom {}))
+(def file-path "data.json")
+
+(defn add-Course
```
`add-course`, not `add-Course` (even if `Course` is a Java type or Protocol... though we can debate this later 😃 )

deobald commented on src/database/core.clj:
```diff
@@ -0,0 +1,23 @@
+
+(defn main
+    []
+    (if (.exists (io/as-file file-path))
```
Try to isolate all Java interop into a separate set of namespaces (or just one ns), eventually.

deobald commented on src/database/core.clj:
```diff
@@ -0,0 +1,23 @@
+    [coursename start duration]
+    (let [this-course {"start" start,
+                      "duration" duration,
+                      "students" {}}]
```
This may be explained later but I'm in a rush... `students` isn't a vector? Why a map?

deobald commented on src/database/core.clj:
```diff
@@ -0,0 +1,23 @@
+
+(defn add-Course
+    [coursename start duration]
+    (let [this-course {"start" start,
```
Keys of maps should (almost) always be keywords. Definitely prefer keywords to strings.

deobald commented on src/gui/core.clj:
```diff
@@ -0,0 +1,59 @@
+
+
+(defn main-screen-frame []
+    (config! f :title "Vipasana"
```
Usually spelled "vipassana" 😺

deobald commented on src/gui/core.clj:
```diff
@@ -0,0 +1,59 @@
+          start-date  (:start-date data)
+          duration    (:duration data)]
+     (cond
+        (or (empty? course-name) (empty? start-date) (empty? duration)) (alert "Please enter all the fields")
```
Eventually (not necessary now), move all validation-related things into their own namespace, including warning/error messages.

deobald commented on src/gui/core.clj:
```diff
@@ -0,0 +1,59 @@
+      :items [
+        [nil :fill :both :insets (java.awt.Insets. 5 5 5 5) :gridx 0 :gridy 0]
+
+        [(label :text "Course Name:" :halign :center :font {:size 20})]
```
Eventually (not now), move all user-facing text into a single ns so it is easy to translate/isolate.

deobald commented on src/validate/core.clj:
```diff
@@ -0,0 +1,12 @@
+(ns validate.core
+  (:gen-class))
```
Nice :)

deobald commented on test/gui/core_test.clj:
```diff
@@ -0,0 +1,7 @@
+  (:require [clojure.test :refer :all]
+            [gui.core :refer :all]))
+
+(deftest a-test
```
There are 2 very easy tests you could write now and a few other (harder) tests that you could try to write, if you want. Ask the group what they think these are and how you might go about writing them.

Do write at least the two easy tests before merging. Otherwise, looks good!

deobald reviewed (changes requested):
sorry! this was the review I forgot to submit,

Commit:
Code refactoring for code review
  - Change add-Course to add-course
  - Change atom name: course to all-course-data

Commit:
Add apis to check if course already exists

Commit:
Change design of add-course screen from form layout to vertical layout

Commit:
Change frame size and disable frame resizing

Commit:
Add indicator to duration field

Commit:
Move empty validations to different namespace

mayankpahwa commented:
@deobald Merging this PR. 

To Do:
  - Convert key of maps to keywords instead of strings(as per the discussion yesterday)
  - Write tests
